lib/iaas/GcpClient.js

```diff
@@ -0,0 +1,225 @@
+    return Promise.try(() =>
+      this.storageClient
+      .bucket(container)
+      .getFiles(queryOptions)
+      .then(results => {
+        const resultFiles = results[0];
+        const files = [];
+        _.each(resultFiles, file => files.push(_
+          .chain(file)
+          .pick('name')
+          .value()
+        ));
+        return files;
+      })
+    );
+  }
+
+  remove(container, file) {
+    if (arguments.length < 2) {
+      file = container;
+      container = this.containerName;
+    }
+    logger.debug(`Deleting file ${file} in container ${container} `);
+    return Promise.try(() =>
+        this.storageClient
+        .bucket(container)
+        .file(file)
+        .delete()
+      )
+      .catchThrow(BaseCloudClient.providerErrorTypes.Unauthorized,
+        new Unauthorized(`Authorization at google cloud storage provider failed while deleting blob ${file} in container ${container}`))
+      .catchThrow(BaseCloudClient.providerErrorTypes.Forbidden,
+        new Forbidden(`Authentication at google cloud storage provider failed while deleting blob ${file} in container ${container}`))
+      .catchThrow(BaseCloudClient.providerErrorTypes.NotFound,
+        new NotFound(`Object '${file}' not found while deleting in container ${container}`));
+  }
+
+  download(options) {
+    return utils.streamToPromise(this.storageClient
+        .bucket(options.container)
+        .file(options.remote)
+        .createReadStream())
+      .catchThrow(BaseCloudClient.providerErrorTypes.NotFound, new NotFound(`Object '${options.remote}' not found`));
+  }
+
+  upload(options, buffer) {
+    return new Promise((resolve, reject) => {
+      function cleanup() {
+        uploadedStream.removeListener('finish', onfinish);
+        uploadedStream.removeListener('error', onerror);
+      }
+
+      function onerror(err) {
+        cleanup();
+        reject(err);
+      }
+
+      function onfinish() {
+        cleanup();
+        resolve(JSON.parse(buffer));
+      }
+      const uploadedStream = this.storageClient
+        .bucket(options.container)
+        .file(options.remote)
+        .createWriteStream();
+      uploadedStream.on('error', onerror);
+      uploadedStream.on('finish', onfinish);
+      uploadedStream.end(buffer);
+    });
+  }
+
+  uploadJson(container, file, data) {
+    if (arguments.length < 3) {
+      data = file;
+      file = container;
+      container = this.containerName;
+    }
+    return this
+      .upload({
+        container: container,
+        remote: file
+      }, new Buffer(JSON.stringify(data, null, 2), 'utf8'));
+  }
+
+  downloadJson(container, file) {
+    if (arguments.length < 2) {
+      file = container;
+      container = this.containerName;
+    }
+    return this
+      .download({
+        container: container,
+        remote: file
+      })
+      .then((data) => JSON.parse(data))
+      .catchThrow(SyntaxError, new NotFound(`Object '${file}' not found`));
```

Comment:
is syntax error to be treated as 404 not found? or an internal Server Error 500?